q2_coordinates/qtrees.py

```diff
@@ -0,0 +1,217 @@
+        self.children = []
+        self.id = _id
+
+    def get_width(self):
+        return self.width
+
+    def get_height(self):
+        return self.height
+
+    def get_points(self):
+        return self.points
+
+    def get_id(self):
+        return self.id
+
+    def set_id(self, _id):
+        self.id = self.id+_id
+
+
+class QTree():
+    def __init__(self, threshold, data):
+        self.threshold = threshold
+        self.points = [Point(x, y, sample_id) for sample_id, x, y in data]
+        x_max = max(x for _, x, _ in data)
+        y_max = max(y for _, _, y in data)
+        self.root = Node(0, 0, x_max, y_max, self.points, "0")
+
+    def add_point(self, x, y, sample_id):
+        self.points.append(Point(x, y, sample_id))
+
+    def get_points(self):
+        return self.points
+
+    def subdivide(self, threshold):
+        depth = 0
+        node_id = ""
+        bins = []
+        bins = recursive_subdivide(self.root, threshold, depth, node_id, bins)
+        return bins
+
+
+def recursive_subdivide(node, k, depth, node_id, bins):
+    if len(node.points) < k:
+        return
+    elif len(node.points)/k >= len(node.points):
+        raise ValueError("The threshold for subdivision is less than "
+                         "the amount of points, "
+                         "please chose a larger threshold for division")
+    w_ = node.width/2
+    h_ = node.height/2
+    depth += 1
+    nodes = []
+    for i in range(4):
+        # northwest
+        if(i == 0):
+            quad = "1"
+            node.x0 = node.x0
+            node.y0 = node.y0+h_
+        # northeast
+        elif(i == 1):
+            quad = "2"
+            node.x0 = node.x0+w_
+            node.y0 = node.y0
+        # southwest
+        elif(i == 2):
+            quad = "3"
+            node.x0 = node.x0-w_
+            node.y0 = node.y0-h_
+        # southeast
+        elif(i == 3):
+            quad = "4"
+            node.x0 = node.x0+w_
+            node.y0 = node.y0
+
+        p = contains(node.x0, node.y0, w_, h_, node.points)
+        new_id = node_id + quad + "."
+        quad_node = Node(node.x0, node.y0, w_, h_, p, new_id)
+        nodes.append(quad_node)
+
+        for pt in p:
+            bins.append((pt.sample_id, depth, new_id))
+
+        recursive_subdivide(quad_node, k, depth, new_id, bins)
+    node.children = nodes
+    return bins
+
+
+def contains(x, y, w, h, points):
+    pts = []
+    for point in points:
+        if point.x >= x and point.x <= x+w and point.y >= y and point.y <= y+h:
+            pts.append(point)
+    return pts
+
+
+def create_tree(bins, index):
+    df = pd.DataFrame(bins, columns=[index, 'depth', 'lineage'])
+    longest_lineages = []
+    print(df)
+    for sample_id, sample_grp in df.groupby(index):
+        sample_grp_sorted = sample_grp.sort_values('depth', ascending=False)
+        longest_lineages.append(sample_grp_sorted.iloc[0])
+    longest_lineages = pd.DataFrame(longest_lineages)
```

Comment:
This piece of code does exactly the same as in the `create_sample_df` function. Couldn't we pass the lineage df directly to `create_tree` rather than calculate it once again from bins?

Reply:
Done, added several comments to make slightly more readable

Reply:
Thought stuff worked, but turns out my "fix" actually broke construction. Seems obvious to me now, but the inputs to the for loops are different (dataframe needing "chopped" lineages, trees needing whole lineages). I'm still looking into ways that this can be dealt with, but for now I am reverting to og.